test/unit/org/apache/cassandra/index/sai/cql/RandomisedComplexQueryTest.java

```diff
@@ -0,0 +1,451 @@
+            builder.append(Streams.concat(partitionKeys.stream(), clusteringKeys.stream(), normalColumns.stream())
+                                  .map(RandomColumn::name)
+                                  .collect(Collectors.joining(", ")));
+            builder.append(") VALUES (");
+            builder.append(Streams.concat(partitionKeys.stream(), clusteringKeys.stream(), normalColumns.stream())
+                                  .map(RandomColumn::bindMarker)
+                                  .collect(Collectors.joining(", ")));
+            builder.append(")");
+
+            return builder.toString();
+        }
+
+        public List<String> generateIndexStrings()
+        {
+            List<String> indexes = new ArrayList<>();
+            clusteringKeys.stream().map(RandomColumn::toIndexDefinition).forEach(indexes::add);
+            normalColumns.stream().map(RandomColumn::toIndexDefinition).forEach(indexes::add);
+            return indexes;
+        }
+
+        public List<RandomRow> generateDataset()
+        {
+            List<RandomRow> data = new ArrayList<>();
+
+            for (int row = 0; row < CQLTester.getRandom().nextIntBetween(100, 1000); row++)
+            {
+                RandomRow newRow = generateRow();
+                // Remove any duplicate rows - makes it easier to build result set
+                // It may be possible to handle duplicates in the filtering but for the
+                // time being we just get rid of them
+                List<RandomRow> duplicates = data.stream()
+                                                 .filter(r -> {
+                                                    for (int pk = 0; pk < partitionKeys.size(); pk++)
+                                                        if (!r.values.get(pk).equals(newRow.values.get(pk)))
+                                                            return true;
+                                                    return false;
+                                                 })
+                                                 .collect(Collectors.toList());
+                duplicates.stream().forEach(data::remove);
+                data.add(newRow);
+            }
+
+            return data;
+        }
+
+        public RandomQuery generateQuery() throws Throwable
+        {
+            StringBuilder builder = new StringBuilder();
+
+            boolean applyPrecedence = CQLTester.getRandom().nextBoolean();
+
+            List<RandomColumn> allColumns = Lists.newArrayList(clusteringKeys);
+            allColumns.addAll(normalColumns);
+            int numberOfElements = CQLTester.getRandom().nextIntBetween(1, allColumns.size());
+            Set<RandomColumn> columns = new HashSet<>();
+            while (columns.size() < numberOfElements)
+                columns.add(allColumns.get(CQLTester.getRandom().nextIntBetween(0, allColumns.size() - 1)));
+
+            RandomColumn[] columnArray = columns.toArray(new RandomColumn[] {});
```

Comment:
Would it be easier to build a random expression tree then convert to a CQL string?

Reply:
No because I want to be able to generate expressions without specific precedence to test that that is working correctly. E.g. `a = 1 OR b = 1 AND c = 1 OR d = 1`. If I built an expression tree, I wouldn't be able to do that (easily).